src/bentoml/_internal/configuration/containers.py

```diff
@@ -298,68 +98,62 @@ def __init__(
                     split_with_quotes(line, sep="=", quote='"') for line in lines
                 ]
             }
+            # Note that this values will only support latest version of configuration,
+            # as there is no way for us to infer what values user can pass in.
+            # however, if users pass in a version inside this value, we will that to migrate up
+            # if possible
+            if "version" in override_config_map:
+                override_version = override_config_map["version"]
+                logger.debug(
+                    "Found defined 'version=%d' in BENTOML_CONFIG_OPTIONS."
+                    % override_version
+                )
+                migration = getattr(
+                    import_configuration_spec(override_version), "migration", None
+                )
+                # Running migration layer if it exists
+                if migration is not None:
+                    override_config_map = migration(override_config=override_config_map)
+            # Previous behaviour, before configuration versioning.
             try:
-                override_config = unflatten(override_config_map)
+                override = unflatten(override_config_map)
             except ValueError as e:
                 raise BentoMLConfigException(
-                    f"Failed to parse config options from the env var: {e}. \n *** Note: You can use '\"' to quote the key if it contains special characters. ***"
+                    f"Failed to parse config options from the env var:\n{e}.\n*** Note: You can use '\"' to quote the key if it contains special characters. ***"
                 ) from None
-            config_merger.merge(self.config, override_config)
+            config_merger.merge(self.config, override)
 
         if override_config_file is not None or override_config_values is not None:
             self._finalize()
 
         if validate_schema:
             try:
-                SCHEMA.validate(self.config)
-            except SchemaError as e:
+                spec_module.SCHEMA.validate(self.config)
+            except s.SchemaError as e:
                 raise BentoMLConfigException(
-                    "Invalid configuration file was given."
-                ) from e
+                    f"Invalid configuration file was given:\n{e}"
+                ) from None
 
     def _finalize(self):
-        global_runner_cfg = {k: self.config["runners"][k] for k in RUNNER_CFG_KEYS}
-        for key in self.config["runners"]:
-            if key not in RUNNER_CFG_KEYS:
-                runner_cfg = self.config["runners"][key]
+        GLOBAL_RUNNERS_KEY = ["batching", "resources", "logging", "metrics", "timeout"]
```

Comment:
why global here?

Reply:
it is just RUNNER_CFG_KEY. Let me rename this back.